numpy/ma/core.py

```diff
@@ -6300,6 +6306,13 @@ def __deepcopy__(self, memo=None):
         memo[id(self)] = copied
         for (k, v) in self.__dict__.items():
             copied.__dict__[k] = deepcopy(v, memo)
+        # as clearly documented for np.copy(), you need to use
+        # deepcopy() directly for arrays of object type that may
+        # contain compound types--you cannot depend on normal
+        # copy semantics to do the right thing here
+        if self.dtype.type is np.object_:
+            for idx, _ in enumerate(copied):
+                copied[idx] = deepcopy(copied[idx])
```

Comment:
Maybe just call deepcopy on `_data`?  You can guard it by `dtype.hasobject` instead.  If we put it here, that may copy it twice, but maybe that is better than the bug.

I don't trust that the code works right now for 2-D masked arrays.